### PR TITLE
Bitfinex WS data parsing

### PR DESCRIPTION
**Pull request by m-an**
+ base exchanges pushing actions logic

### CODE REVIEW

File: lib/arke/exchange/base.rb [hunk 1]
```diff
@@ -6,5 +8,9 @@ def initialize(config, strategy)
 
     def call(action)
     end
+
+    def push(action, data)
+      @strategy.push(Arke::Action.new(action, data))
```
Comment (dnfd) [inline lib/arke/exchange/base.rb]: what is data? Action has params for it's data I guess

File: lib/arke/strategy/copy.rb [hunk 2]
```diff
@@ -1,7 +1,12 @@
+require 'action'
+
 module Arke::Strategy
   class Copy < Base
+    attr_accessor :market
+
     def initialize(config)
       super
+      @market = config['market']
```
Comment (dnfd) [inline lib/arke/strategy/copy.rb]: config['pair'], I guess

**Review by dnfd — DISMISSED** [review]

File: lib/arke/exchange/base.rb (hunk 1 above)
Comment (m-an) [reply lib/arke/exchange/base.rb]: It's a method for creating an `Arke::Action` from your data

File: lib/arke/exchange/bitfinex.rb [hunk 3]
```diff
@@ -1,15 +1,19 @@
+# frozen_string_literal: true
+
+require 'order'
```
Comment (mod) [inline lib/arke/exchange/bitfinex.rb:4]: remove order

File: lib/arke/exchange/bitfinex.rb [hunk 4]
```diff
@@ -1,15 +1,19 @@
...
       # it solves a problem with queue.pop, websocket will push messages
       # exchange handles parsing and subscription logic
-      Arke::Log.debug "Bitfinex processes action #{action}"
-
-      if action.type == :websocket
-        # parsing and sending to strategy
-        # @strategy.push_order
-      end
+      Arke::Log.debug 'Bitfinex processes action'
+      Arke::Log.debug action.inspect
```
Comment (mod) [inline lib/arke/exchange/bitfinex.rb]: action has TO_S

File: lib/arke/exchange/bitfinex.rb [hunk 5]
```diff
@@ -22,14 +24,43 @@ def process_channel_message(msg)
...
+    end
+
+    def parse_ws_data(data)
+      data[2].positive? ? build_buy_order(data) : build_sell_order(data)
+    end
+
+    def push(action, order)
+      @strategy.push(Arke::Action.new(action, order))
+    end
+
```
Comment (mod) [inline lib/arke/exchange/bitfinex.rb:45]: todo: avoid duplicating code, also order can be smarter and use the same convention

File: lib/arke/order.rb [hunk 6]
```diff
@@ -3,16 +3,12 @@ class Order
 
     attr :id, :market, :price, :amount, :side
 
-    def initialize(id, market, price, amount)
+    def initialize(id, market, price, amount, side)
```
Comment (mod) [inline lib/arke/order.rb:6]: let's put side after market

File: lib/arke/strategy/base.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+module Arke::Strategy
+  # Base class for all strategies
+  class Base
+    attr_reader :pair
+
+    def initialize(config)
+      @sources = []
+      @target = nil
+      @pair = config['pair']
+      @order_queue = Queue.new
```
Comment (mod) [inline lib/arke/strategy/base.rb:13]: remove the queue

**Review by mod — APPROVED** [review]

File: lib/arke/exchange/bitfinex.rb (hunk 5 above)
Comment (m-an) [reply lib/arke/exchange/bitfinex.rb:45]: I guess we better move it to Base class

**Pull request merged**